### PR TITLE
Implementation of a specific  e-field distortion correction

### PR DESCRIPTION
**Pull request by Yassid**

### CODE REVIEW

File: AtTools/AtEDistortionModel.cxx [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+   auto [zcorr, radcorr, tracorr] = GetCorrectionFactors((Int_t)inputRad, (Int_t)inputZ);
+
+   Double_t outputRad = TMath::Sqrt(TMath::Power(inputRad + radcorr, 2) + TMath::Power(tracorr, 2));
+   Double_t outputAng = inputAng + TMath::ATan2(tracorr, inputRad + radcorr);
+
+   XYZPoint output;
+   output.SetX(outputRad * TMath::Cos(outputAng));
+   output.SetY(outputRad * TMath::Sin(outputAng));
+   output.SetZ(input.Z() - zcorr);
```
Comment (anthoak13) [inline AtTools/AtEDistortionModel.cxx:58]: Can use GenVector again to do stuff like:
```
output.SetRho(outputRad)
output.SetPhi(outputAng)
```

File: AtTools/AtEDistortionModel.cxx [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+   LOG(debug) << " Output point : " << output.X() << " " << output.Y() << " " << output.Z() << "\n";
+
+   return output;
+}
+
+XYZPoint AtEDistortionModel::ApplySpaceCharge(const XYZPoint &input)
+{
+   XYZPoint p;
```
Comment (anthoak13) [inline AtTools/AtEDistortionModel.cxx:67]: I think you could shorten this to `return {};`

File: AtTools/AtEDistortionModel.cxx [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+      std::cerr << " AtEDistortionModel::SetCorrectionMaps - Error! Missing LUT "
+                << "\n";
+      std::cerr << zlut << "\n";
+      std::cerr << radlut << "\n";
+      std::cerr << tralut << "\n";
+      return kFALSE;
+   }
+
+   return kTRUE;
```
Comment (anthoak13) [inline AtTools/AtEDistortionModel.cxx:99]: unreachable code, return statements in both branches of if.

File: AtTools/AtEDistortionModel.cxx [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+std::tuple<Double_t, Double_t, Double_t>
+AtEDistortionModel::GetCorrectionFactors(const Int_t &radius, const Int_t &zpos)
```
Comment (anthoak13) [inline AtTools/AtEDistortionModel.cxx:103]: Fine to take references, but is the same as just using an `Int_t radius` (addresses and ints are the same size).

This also doesn't matter, and is just a style thing, but you can just pass around EFieldCorr objects to avoid converting between those and tuples in the code.

File: AtTools/AtEDistortionModel.cxx [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+      Double_t value;
+      std::istringstream buffer(line);
+      Int_t cnt = 0;
+
+      buffer >> radius;
+
+      while (buffer >> value) {
+
+         EFieldMapRef ref(radius, cnt);
+         map->emplace(ref, value);
```
Comment (anthoak13) [inline AtTools/AtEDistortionModel.cxx:140]: Could probably replace with `map->emplace({radius, cnt}, value);` Little shorted and imo just as readable.

File: AtTools/AtEDistortionModel.cxx [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+   std::cout << " Tra map size : " << tramap->size() << "\n";
+
+   // NB: This is not a necessary condition but it works for the present case
+   assert((zmap->size() == radmap->size()) && (radmap->size() == tramap->size()));
+
+   for (const auto &[key, zcorr] : *zmap) {
+      auto radcorr = radmap->at(key);
+      auto tracorr = tramap->at(key);
+      EFieldCorr corr(zcorr, radcorr, tracorr);
+      fDistortionMap->emplace(key, corr);
```
Comment (anthoak13) [inline AtTools/AtEDistortionModel.cxx:160]: Similar to above comment, can use aggregate initialization (https://en.cppreference.com/w/cpp/language/aggregate_initialization).

File: AtTools/AtEDistortionModel.cxx [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+      }
+   }
+}
+
+void AtEDistortionModel::SetGlobalMap(BufferMap *zmap, BufferMap *radmap, BufferMap *tramap)
+{
+
+   std::cout << " Z map size : " << zmap->size() << "\n";
+   std::cout << " Rad map size : " << radmap->size() << "\n";
+   std::cout << " Tra map size : " << tramap->size() << "\n";
```
Comment (anthoak13) [inline AtTools/AtEDistortionModel.cxx:151]: couts should maybe be LOG(info)s?

File: AtTools/AtEDistortionModel.cxx [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+   EFieldMapRef hashRef(200, 200);
+
+   std::cout << " Global map size : " << fDistortionMap->size() << "\n";
+   std::cout << " Hash value for (200,200)     : " << hashT(hashRef) << "\n";
+}
+
+void AtEDistortionModel::LoadParameters(AtDigiPar *par)
+{
+   if (par == nullptr)
+      return;
```
Comment (anthoak13) [inline AtTools/AtEDistortionModel.cxx:174]: Doesn't actually do anything, can just be `{}`

File: AtTools/AtEDistortionModel.h [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+class AtDigiPar;
+/**
+ * @brief Model for E-field distortion correction
+ *
+ *
+ *
+ *     The implemented correction is applied to every hit
+ *     Radial, transversal and longitudinal (z) corrections as a function of radial distance and z (map keys).
+ *     The Class EField mapping contains the correction values
+ */
```
Comment (anthoak13) [inline AtTools/AtEDistortionModel.h:21]: This doxygen comment will refer just to the struct EFieldCorr, which I don't think is the intention? It seems to be describing the entire model so should be above the actual model class.

File: AtTools/AtEDistortionModel.h [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+   inline void SetRadCorr(Double_t radcorr) { _radcorr = radcorr; }
+   inline void SetTraCorr(Double_t tracorr) { _tracorr = tracorr; }
+   inline Double_t GetZCorr() { return _zcorr; }
+   inline Double_t GetRadCorr() { return _radcorr; }
+   inline Double_t GetTraCorr() { return _tracorr; }
+
+   Double_t _zcorr;
+   Double_t _radcorr;
+   Double_t _tracorr;
+};
```
Comment (anthoak13) [inline AtTools/AtEDistortionModel.h:38]: This has too much machinery around it that will keep users from using modern c++ features with structs, many of which are disabled when user-provided constructors (or default init values) are added and they are treated like classes instead of like structs. Should probably just be 
```
struct EFieldCorr {
Double_t zcorr;
Double_t radcorr;
Double_t tracorr;
};
```
(See AtPadReference for example)

File: AtTools/AtEDistortionModel.h [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+namespace std {
+template <>
+struct hash<EFieldMapRef> {
+   inline size_t operator()(const EFieldMapRef &x) const { return x._rad + x._zpos * 10000; }
+};
+} // namespace std
+
+using eDistortionMap = std::unordered_map<EFieldMapRef, EFieldCorr>; // Global LUT
+using BufferMap = std::unordered_map<EFieldMapRef, Double_t>;        //! To parse individual LUT
```
Comment (anthoak13) [inline AtTools/AtEDistortionModel.h:61]: I would put these in the AtDistortionModel class so they don't pollute the global namespace.

**Review by anthoak13 — CHANGES_REQUESTED** [review]
There are a couple of style things to match the coding standards, and I think it would be useful to fully embrace using structs rather than half class/half structs for the internal data structures used in the distortion model.

File: AtGenerators/AtTPCIonDecay.cxx [hunk 12]
```diff
@@ -256,9 +256,8 @@ Bool_t AtTPCIonDecay::ReadEvent(FairPrimaryGenerator *primGen)
       if (s > pow(M_tot, 2)) {
          // if(ExEject*1000.0>fSepEne){
          fIsDecay = kTRUE;
-         event1.SetDecay(fEnergyImpulsionLab_Total, fMult.at(Case), mass_1);
-         // Double_t weight1 = event1.Generate();
-
+         Bool_t decay = event1.SetDecay(fEnergyImpulsionLab_Total, fMult.at(Case), mass_1);
+         Double_t weight1 = event1.Generate();
```
[the remark this answers is not in the document]
Comment (Yassid) [reply AtGenerators/AtTPCIonDecay.cxx:260]: This is not part of the code I have modified

File: AtTools/AtEDistortionModel.cxx (hunk 5 above)
Comment (Yassid) [reply AtTools/AtEDistortionModel.cxx:140]: Wasnt able to compile it

Comment (anthoak13) [reply AtTools/AtEDistortionModel.cxx]: For this to compiled the suggested changes to the struct had to be made to so it could use aggregate initialization https://en.cppreference.com/w/cpp/language/aggregate_initialization

Comment (Yassid) [reply AtTools/AtEDistortionModel.cxx]: Thats what I actually did, but it didnt like it
141:36: error: no matching function for call to ‘std::unordered_map<EFieldMapRef, double>::emplace(<brace-enclosed initializer list>, Double_t&)’
  141 |     map->emplace({radius,cnt},value);

Comment (anthoak13) [reply AtTools/AtEDistortionModel.cxx]: Someday I'll figure out why that is. I expected it to work like this `fAtMapPtr->AddAuxPad({10, 0, 0, 0}, "MCP_US");` but something must be different with the map emplace function signature. My guess not testing is emplace takes an r value, and it can't bind to whatever is actually being returned by the `{}` while in a the AddAuxPad it can bind to the const reference used there.
